Fips140-3 Configure tests

abutch3r opened the pull request:
Add Tests for checking that the scripts correctly configure themselves for FIPS140-3. This bucket applies to both IBM Semeru 11, 17, 21, 25+ and IBM SDK 8.

The tests check to ensure we are in a compatible environment based on the Liberty requirements for FIPS140-3.

The tests cover setting the JVM options, Environment variables manually in the env or via the correct `.env` files.
It covers the securityUtility configureFIPS task scenarios cover enablement at an install, server and client level

Update LibertyClient to support the supplying of environment variables as we do for LibertyServer so that the client can have its env set

Update LibertyClient to more accurately track its state in relation to when it is stoped

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [ ] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [ ] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

abutch3r committed:
Add tests for FIPS 140-3 Enablement and securityUtility

Add tests that just cover the server and client scripts taking the ENABLE_FIPS140_3 envvar and use it to ensure we can isolate those scripts behaviour.
Add tests that use securityUtility to set various enablement configurations and scenarios. includes starting servers and clients for end-to-end scenarios

Applies to both Semeru versions that support Liberty's minimum FIPS levels and IBM SDK 8

Checks within the test ensure we have a valid environment to run the tests. AlwaysPassesTest is included from when base requirements are not met

Update LibertyClient to support the ability to add the ability to addEnvVars as we do for LibertyServer.
Fix LibertyClient tracking of isStarted behaviour. isStopped was never used, but isStarted was never set back to false, once the client had stopped.

abutch3r commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=f79dd7e6-9dc2-4a59-b5d9-589d07714853) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jimmy1wu commented on dev/io.openliberty.security.fips.internal_fat/fat/src/io/openliberty/security/fips/fat/tests/fips1403/client/FIPS140_3_EmptyEnvClientTest.java:
```diff
@@ -0,0 +1,49 @@
+@SkipIfSysProp({SkipIfSysProp.OS_ZOS, SkipIfSysProp.OS_IBMI, SkipIfSysProp.OS_ISERIES})
+public class FIPS140_3_EmptyEnvClientTest {
+
+    private final static String CLIENT_NAME="BasicClinet";
```
there's a typo in the client name

jimmy1wu commented on dev/io.openliberty.security.fips.internal_fat/fat/src/io/openliberty/security/fips/fat/tests/fips1403/security/utility/FIPS1403SecurityUtilityTests.java:
```diff
@@ -0,0 +1,210 @@
+        }
+        if (fipsDefaultProfileFile.exists()) fipsDefaultProfileFile.delete();
+        if (serverEnv.exists()) serverEnv.delete();
+        if (fipsServerProfileFile.exists()) fipsServerProfileFile.delete();
```
there is also a `default.env.backup` / `server.env.backup` / `client.env.backup` file that's created whenever we modify the env files.

while non-functional, i think we should cleanup those files as well. maybe just the `default.env.backup` at the very least since that's in the etc directory.

jimmy1wu commented on dev/io.openliberty.security.fips.internal_fat/fat/src/io/openliberty/security/fips/fat/tests/fips1403/server/FIPS1403ServerTest.java:
```diff
@@ -0,0 +1,198 @@
+    @Test
+    public void fips140_3DirectorySpaceNoQuotesTest() throws Exception {
+        assumeThat(GLOBAL_FIPS, is(false));
+        assumeThat(server.getMachine().getOperatingSystem(), is(OperatingSystem.WINDOWS));
```
i think we can take out this assumes windows line, since space with no quotes (and no variable expansion) should work on non-windows as well.

jimmy1wu commented on dev/io.openliberty.security.fips.internal_fat/fat/src/io/openliberty/security/fips/fat/tests/fips1403/server/FIPS1403ServerTest.java:
```diff
@@ -0,0 +1,198 @@
+    @Test
+    public void fips140_3DirectorySpaceQuotesTest() throws Exception {
+        assumeThat(GLOBAL_FIPS, is(false));
+        // assumeThat(server.getMachine().getOperatingSystem(), not(OperatingSystem.WINDOWS));
```
should this be uncommented out? since there isn't variable expansion on windows and the similar `fips140_3DirectorySpaceSlashTest` test assumes non-windows as well.

jimmy1wu reviewed (commented):
hi alex, thanks for the tests.

they look good overall with just some minor comments.

abutch3r commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9b647724-539d-4e0d-9bb0-9c42f7b12ef7) - :warning: completed with warnings.)

_Note: Target locations of links might be accessible only to IBM employees._

abutch3r committed:
Update FIPS Tests

Update Server and Client classes to detect that the test is enabling FIPS and global fips is enabled, to not use the global fips settings

Fix some Java 8 and Semeru bugs

abutch3r commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=adc3ae5b-e857-434b-86df-17c9aafb0627) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._
